test/sparse.jl

```diff
@@ -18,33 +18,119 @@
+            y = f(x)
+            Js = jacobian_sparsity(f, x, sd)
+            @test Js isa AbstractMatrix
+            @test size(Js) == (length(y), length(x))
+            @test all(isone, Js)
+        end
+
+        for x in (rand(2), rand(2, 3)), y in (rand(5), rand(5, 6))
+            Js = jacobian_sparsity(f_jac!, y, x, sd)
+            @test Js isa AbstractMatrix
+            @test size(Js) == (length(y), length(x))
+            @test all(isone, Js)
+        end
+
+        for x in (rand(2), rand(2, 3))
+            Hs = hessian_sparsity(f_hess, x, sd)
+            @test Hs isa AbstractMatrix
+            @test size(Hs) == (length(x), length(x))
+            @test all(isone, Hs)
+        end
     end
 
-    for x in (rand(2), rand(2, 3)), y in (rand(5), rand(5, 6))
-        Js = jacobian_sparsity(f_jac!, y, x, sd)
-        @test Js isa AbstractMatrix
-        @test size(Js) == (length(y), length(x))
-        @test all(isone, Js)
+    @testset "KnownJacobianSparsityDetector" begin
+        @testset "Out-of-place functions" begin
+            for sx in ((2,), (2, 3)), f in (f_jac1, f_jac2)
+                x = rand(sx...)
+                nx = length(x)
+                Jref = rand(Bool, 2 * nx, nx)
+                sd = KnownJacobianSparsityDetector(Jref)
+                Js = jacobian_sparsity(f, x, sd)
+                @test Js isa AbstractMatrix
+                @test size(Js) == (2 * nx, nx)
+                @test Js === Jref
+            end
+        end
+        @testset "In-place functions" begin
+            for sx in ((2,), (2, 3)), sy in ((5,), (5, 6))
+                x, y = rand(sx...), rand(sy...)
+                nx, ny = length(x), length(y)
+                Jref = rand(Bool, ny, nx)
+                sd = KnownJacobianSparsityDetector(Jref)
+                Js = jacobian_sparsity(f_jac!, y, x, sd)
+                @test Js isa AbstractMatrix
+                @test size(Js) == (ny, nx)
+                @test Js === Jref
+            end
+        end
+        @testset "Exceptions: hessian_sparsity not supported" begin
+            for sx in ((2,), (2, 3))
+                x = rand(sx...)
+                nx = length(x)
+                Href = rand(Bool, nx, nx)
+                sd = KnownJacobianSparsityDetector(Href)
+                @test_throws ArgumentError hessian_sparsity(f_hess, x, sd)
+            end
+        end
+        @testset "Exceptions: DimensionMismatch" begin
+            sd = KnownJacobianSparsityDetector(rand(Bool, 6, 7)) # wrong Jacobian size
+            for x in (rand(2), rand(2, 3)), f in (f_jac1, f_jac2)
+                @test_throws DimensionMismatch jacobian_sparsity(f, x, sd)
+            end
+            for x in (rand(2), rand(2, 3)), y in (rand(5), rand(5, 6))
+                @test_throws DimensionMismatch jacobian_sparsity(f_jac!, y, x, sd)
+            end
+        end
     end
+    @testset "KnownHessianSparsityDetector" begin
+        for sx in ((2,), (2, 3))
```

Comment:
let's put this in a test set and then we'll be good to go! (sorry I didn't spot it at first)